src/blog/2018-09-24/mutation-switching.pug

```diff
@@ -0,0 +1,116 @@
+        Previously a very trivial example was used. With this statement, it was easy to mutate using mutation switching. 
+        But what if the statements get more complex? How can complex code be avoided? Because the more complex the code is the higher change compilation errors occur. 
+        For example
+        ```scala
+        def isEven(number: Int): Boolean = number % 2 == 0
+        def isOdd(number: Int): Boolean = !isEven(number)
+
+        val numbers: List[Int] = 1 to 100 toList
+
+        def filterOddAnd(specialNumber: Int): List[Int] = {
+          numbers
+            .filter(isOdd)
+            .filterNot(_.equals(specialNumber))
+        }
+        ```
+        With this code base `filter` and `filterNot` could be mutated to there counterparts. 
+        This would give us the following code base if we implement the pattern match at the direct position. 
+        ```scala
+        def isEven(number: Int): Boolean = number % 2 == 0
+        def isOdd(number: Int): Boolean = !isEven(number)
+
+        val numbers: List[Int] = 1 to 100 toList
+
+        def filterOddAnd(specialNumber: Int): List[Int] = {
+          numbers.(sys.env.get("ACTIVE_MUTATION") match {
+            case Some("1") => .filter(isOdd)
+            case _ => .filterNot(isOdd)
+          }).
+          (sys.env.get("ACTIVE_MUTATION") match {
+            case Some("1") =>  .filterNot(_.equals(specialNumber))
+            case _ =>  .filter(_.equals(specialNumber))
+          })
+        }
+        ```
+        As you can see it's already a lot more complicated and the code doesn't even compile! 
+        To reduce a lot of the complexity and make the code actually compile we need to take a deeper look in the abstract syntax tree 
+        and find the corresponding top statements which is `numbers` in our case. If we implement mutation switching using top statements
+        we can reduce the code complexity and generate the following code. 
+        ```Scala
+        def isEven(number: Int): Boolean = number % 2 == 0
+        def isOdd(number: Int): Boolean = !isEven(number)
+ 
+        val numbers: List[Int] = 1 to 100 toList
+ 
+        def filterOddAnd(specialNumber: Int): List[Int] = {
+          sys.env.get("ACTIVE_MUTATION") match {
+            case Some("0") => numbers.filterNot(isOdd).filter(_.equals(specialNumber))
+            case Some("1") => numbers.filter(isOdd).filter(_.equals(specialNumber))
+            case _                => numbers.filter(isOdd).filterNot(_.equals(specialNumber))
+          }
+        }
+        ```
+        This enables us to get clean, readable pattern matches and avoid compilation errors like we've seen before. 
+        
+        ## Conclusion
+        With the combination of mutation switching and traversing back to the top statements Stryker4s is able to apply mutations to the codebase 
```

Comment:
What is the AST term for a "top statement"? Can we use that term instead of "top statement"?

Reply:
I think its just statement i will change it to parent statement i guess thats clearer